File: libraries/sensors/sensors.cpp
#include "sensors.h"

//Constructor
sensors::sensors() {
  //Make sure this is the same as the modeling matrix-1
  NUMVARS = 29;
  sense_matrix.zeros(NUMVARS,1,"Full State From Sensors");
  //sense_matrix_dot.zeros(15,1,"Full Statedot From Sensors");
  //Set names of headers
  headernames = (char**)malloc(NUMVARS*sizeof(char*));
  headernames[0] = "Sense X(m)";
  headernames[1] = "Sense Y(m)";
  headernames[2] = "Sense Z(m)";
  headernames[3] = "Sense Roll (deg)";
  headernames[4] = "Sense Pitch (deg)";
  headernames[5] = "Sense Compass (deg)";
  headernames[6] = "Sense U(m/s)";
  headernames[7] = "Sense V(m/s)";
  headernames[8] = "Sense W(m/s)";
  headernames[9] = "Sense P(rad/s)";
  headernames[10] = "Sense Q(rad/s)";
  headernames[11] = "Sense R(rad/s)";
  headernames[12] = "Sense Mx(Gauss)";
  headernames[13] = "Sense My(Gauss)";
  headernames[14] = "Sense Mz(Gauss)";
  headernames[15] = "Sense GPS Latitude (deg)";
  headernames[16] = "Sense GPS Longitude (deg)";
  headernames[17] = "Sense GPS Altitude (m)";
  headernames[18] = "Sense GPS Heading (deg)";
  headernames[19] = "Sense IMU Heading (deg)";
  headernames[20] = "Sense Analog 1 (V)";
  headernames[21] = "Sense Analog 2 (V)";
  headernames[22] = "Sense Analog 3 (V)";
  headernames[23] = "Sense Analog 4 (V)";
  headernames[24] = "Sense Analog 5 (V)";
  headernames[25] = "Sense Analog 6 (V)";
  headernames[26] = "Sense Pressure (Pa)";
  headernames[27] = "Sense Pressure Altitude (m)";
  headernames[28] = "Sense Temperature (C)";
}

//Get numvars
int sensors::getNumVars() {
  return NUMVARS;
}

void sensors::setBias(MATLAB bias_matrix,double bias,double std) {
  for (int i = 1;i<4;i++) {
    double val = bias + randnum(-std,std);
    bias_matrix.set(i,1,val);
  }
}

double sensors::pollute(double bias,double noise) {
  return bias + randnum(-noise,noise);
}

void sensors::init(MATLAB in_configuration_matrix,MATLAB in_simulation_matrix) {
  //Get Rates of All Sensors
  GPS_RATE = in_configuration_matrix.get(5,1);
  IMU_RATE = in_configuration_matrix.get(6,1);
  ANALOG_RATE = in_configuration_matrix.get(7,1);
  IBARO = in_configuration_matrix.get(8,1);

  //Pick the IMU you want to use
  //0 = MPU9250
  //1 = LSM9DS1
  int IMUTYPE = in_configuration_matrix.get(9,1);
  #ifndef HIL
  //Only Initialize IMU if we're not in HIL mode
  initIMU(IMUTYPE);
  #endif

  //Set the Filter Constant
  orientation.FilterConstant = in_configuration_matrix.get(10,1);

  //Initialize GPS
  satellites.init();

  //Initialize Barometer
  atm.init();

  //Initialize q0123 and ptp
  q0123.zeros(4,1,"Sense Quaternions");
  ptp.zeros(3,1,"Sense Roll Pitch Yaw");

  //Sensor Errors
  IERROR = in_simulation_matrix.get(21,1);
  //POSITION
  bias_pos = in_simulation_matrix.get(22,1);
  std_pos = in_simulation_matrix.get(23,1);
  noise_pos = in_simulation_matrix.get(24,1);
  bias_pos_matrix.zeros(3,1,"Bias Position");
  setBias(bias_pos_matrix,bias_pos,std_pos);
  //ANGLE
  bias_angle = in_simulation_matrix.get(25,1);
  std_angle = in_simulation_matrix.get(26,1);
  noise_angle = in_simulation_matrix.get(27,1);
  bias_angle_matrix.zeros(3,1,"Bias Angle");
  setBias(bias_angle_matrix,bias_angle,std_angle);
  //PRESSURE
  bias_pressure = in_simulation_matrix.get(28,1);
  std_pressure = in_simulation_matrix.get(29,1);
  noise_pressure = in_simulation_matrix.get(30,1);
  bias_pressure_val = bias_pressure + randnum(-std_pressure,std_pressure);
  //GYRO
  bias_gyro = in_simulation_matrix.get(31,1);
  std_gyro = in_simulation_matrix.get(32,1);
  noise_gyro = in_simulation_matrix.get(33,1);
  bias_gyro_matrix.zeros(3,1,"Bias Gyro");
  setBias(bias_gyro_matrix,bias_gyro,std_gyro);  
  //MAG
  bias_mag = in_simulation_matrix.get(34,1);
  std_mag = in_simulation_matrix.get(35,1);
  noise_mag = in_simulation_matrix.get(36,1);
  bias_mag_matrix.zeros(3,1,"Bias Mag");
  setBias(bias_mag_matrix,bias_mag,std_mag);
}

///Initialize the IMU
void sensors::initIMU(int sensor_type) {
  //sensor_type == 0 -> MPU9250
  //sensor_type == 1 -> LSM9DS1
  orientation.init(sensor_type);
}
void sensors::sendXYZ2GPS(double currentTime,double X,double Y,double Z) {
  //If it's not time to update GPS
  if (currentTime < nextGPStime) {
    //return prematurely
    return;
  }

  //Add Errors if present
  if (IERROR) {
    X += pollute(bias_pos_matrix.get(1,1),noise_pos);
    Y += pollute(bias_pos_matrix.get(2,1),noise_pos);
    Z += pollute(bias_pos_matrix.get(3,1),noise_pos);
  }
  satellites.X = X;
  satellites.Y = Y;
  satellites.Z = Z;
  //We need to reset the GPS values if we haven't gotten a valid coorinate yet
  satellites.reset();
  //printf("%lf %lf %lf \n",sense.satellites.X,sense.satellites.Y,sense.satellites.Z);
  //PAUSE();
}

void sensors::send(double currentTime,MATLAB model_matrix) {
  //X,Y,Z
  double X = model_matrix.get(1,1);
  double Y = model_matrix.get(2,1);
  double Z = model_matrix.get(3,1);
  //Grab Zpressure before Z is polluted
  double Zpressure = Z;

  //The pollution of XYZ is inside this routine
  sendXYZ2GPS(currentTime,X,Y,Z);

  //Quaternions
  //Convert the quaternions to Euler Angles if ERRORS present
  double q0,q1,q2,q3;
  if (IERROR) {
    q0123.vecset(1,4,model_matrix,4);
    //q0123.disp();
    ptp.quat2euler(q0123);
    //ptp.disp();
    double roll = ptp.get(1,1)*180.0/PI;
    double pitch = ptp.get(2,1)*180.0/PI;
    double yaw = ptp.get(3,1)*180.0/PI;
    roll += pollute(bias_angle_matrix.get(1,1),noise_angle);
    pitch += pollute(bias_angle_matrix.get(2,1),noise_angle);
    yaw += pollute(bias_angle_matrix.get(3,1),noise_angle);
    ptp.set(1,1,roll*PI/180.0);
    ptp.set(2,1,pitch*PI/180.0);
    ptp.set(3,1,yaw*PI/180.0);
    q0123.euler2quat(ptp);
    q0 = q0123.get(1,1);
    q1 = q0123.get(2,1);
    q2 = q0123.get(3,1);
    q3 = q0123.get(4,1);
  } else {
    //model_matrix.disp();
    q0 = model_matrix.get(4,1);
    q1 = model_matrix.get(5,1);
    q2 = model_matrix.get(6,1);
    q3 = model_matrix.get(7,1);
  }
  //This converts doubles to floats because the AHRS filter uses floats
  //I went ahead and changed these to doubles though so that we're more accurate
  //during simulation. This may break auto mode so be sure to revisit
  orientation.ahrs.q0 = q0;
  orientation.ahrs.q1 = q1;
  orientation.ahrs.q2 = q2;
  orientation.ahrs.q3 = q3;
  //printf("send() %lf %lf %lf \n",sense.orientation.roll,sense.orientation.pitch,sense.orientation.yaw);

  //Set gx,gy,gz from sense_matrix for filtering 
  double gy = model_matrix.get(11,1);
  double gx = model_matrix.get(12,1);
  double gz = -model_matrix.get(13,1);
  if (IERROR) {
    //printf("GXYZ Before = %lf %lf %lf \n",gx,gy,gz);
    gx += pollute(bias_gyro_matrix.get(1,1),noise_gyro);
    gy+= pollute(bias_gyro_matrix.get(2,1),noise_gyro);
    gz += pollute(bias_gyro_matrix.get(3,1),noise_gyro);
    //printf("GXYZ After = %lf %lf %lf \n",gx,gy,gz);
  }
  orientation.gx = gx;
  orientation.gy = gy;
  orientation.gz = gz;

  //Pass Z coordinate to barotemp
  if (IERROR) {
    Zpressure += pollute(bias_pressure_val,noise_pressure);
  }
  atm.SendZ(Zpressure);

  //Still need all the other sensor states but not right now

}

void sensors::getCompassHeading() {
  //printf("IMU Yaw = %lf, MAG Yaw = %lf, GPS Heading = %lf \n",orientation.yaw,orientation.magyaw,satellites.heading);
  //For now just pass IMU yaw through to yaw angle
  //compass = orientation.yaw;
  //Or if you want use GPS heading
  //compass = satellites.heading;
  //OR WE CREATE A FUZZY LOGIC FILTER WHERE WE FUSE THE IMU AND THE GPS
  compass = orientation.yaw + heading_offset;
  //Fix wrap between +-180
  if (compass > 180) {
    compass -= 360;
  }
  if (compass < -180) {
    compass += 360;
  }
  //compass = 400;
}

double sensors::getTemperature() {
  return atm.temperature;
}

double sensors::mag_heading(double roll,double pitch,double mx,double my,double mz) {
  printf("RP and Mag: %lf %lf %lf %lf %lf ",roll,pitch,mx,my,mz);
  x_tilt = my*sin(roll*DEG2RAD)*sin(pitch*DEG2RAD) + mx*cos(roll*DEG2RAD) + mz*cos(pitch*DEG2RAD)*sin(roll*DEG2RAD);
  y_tilt = my*cos(pitch*DEG2RAD) - mz*sin(pitch*DEG2RAD);
  printf("Tilt Comp: %lf %lf ",x_tilt,y_tilt);
  if(x_tilt > 0.0){
    Hm = 180. - atan(y_tilt/x_tilt)*RAD2DEG;
  } else if (x_tilt > 0.0 && y_tilt < 0.0){
    Hm = -atan(y_tilt/x_tilt)*RAD2DEG;
  } else if (x_tilt > 0.0 && y_tilt > 0.0) {
    Hm = 360. - atan(y_tilt/x_tilt)*RAD2DEG;
  } else if (x_tilt == 0.0 && y_tilt < 0.0) {
    Hm = 90.;
  } else if (x_tilt == 0.0 && y_tilt > 0.0) {
    Hm = 270.;
  }
  printf("IMU Heading: %lf \n",Hm);
  return Hm;
}

//Polling routine to read all sensors
void sensors::poll(double currentTime,double elapsedTime) {
  //First Analog to Digital Converter
  if (currentTime >= nextANALOGtime) {
    //printf("Reading Analog %lf \n",currentTime);
    analog.get_results();
    nextANALOGtime=currentTime+ANALOG_RATE;
  }
  
  //Then we poll the barometer and temperature sensor (needs current time)
  if (IBARO) {
    //printf("Polling Barometer %lf \n",currentTime);
    atm.poll(currentTime);
  }

  ///Read the IMU (ptp,pqr)
  //IMU must be read as fast as possible due to the elapsedTime
  //and the integrator on board
  orientation.loop(elapsedTime);
  //Maxwell Cobar's mag heading loop 
  //Heading_Mag = mag_heading(orientation.roll,orientation.pitch,orientation.mx,orientation.my,orientation.mz);

  //Read the GPS
  if (currentTime >= nextGPStime) {
    //printf("Polling GPS %lf \n",currentTime);
    satellites.poll(currentTime); //This will compute XYZ as well. For now we are using
    if (currentTime >= nextGPSOffset) {
      heading_offset = satellites.heading - orientation.yaw;
      //printf("Setting Offset = %lf \n",heading_offset);
      nextGPSOffset = currentTime + GPS_RATE;
    }
    //printf("GPS Heading = %lf, Compass Heading = %lf, IMU Heading = %lf Offset = %lf \n",satellites.heading,compass,orientation.yaw,heading_offset); 
    //hardcoded GPS coordinates
    nextGPStime = currentTime + GPS_RATE;
  }
    
  //Then populate the appropriate matrices
  populate(currentTime,elapsedTime);
    
}

void sensors::populate(double currentTime,double elapsedTime) {

  ///////////////////??THEN POPULATE STATE VECTOR////////////////////

  //Initialize everything to -99
  sense_matrix.mult_eq(0);
  sense_matrix.plus_eq(-99);
  //sense_matrix_dot.mult_eq(0);
  //sense_matrix_dot.plus_eq(-99);

  ///XYZ
  sense_matrix.set(1,1,satellites.X);
  sense_matrix.set(2,1,satellites.Y);
  //sense_matrix.set(3,1,(satellites.Z-atm.altitude)/2.0);  //Average GPS and BARO?
  sense_matrix.set(3,1,-atm.altitude); //Let's use Barometer as altitude sensor
  //sense_matrix.set(3,1,-atm.altitude);

  //Roll pitch Yaw
  sense_matrix.set(4,1,orientation.roll);
  sense_matrix.set(5,1,orientation.pitch);

  ///Yaw Angle IS A COMBINATION OF IMU AND GPS
  getCompassHeading();
  sense_matrix.set(6,1,compass);

  //UWV
  //Assume that the vehicle is traveling straight so V and W are zero
  sense_matrix.set(7,1,satellites.speed);
  sense_matrix.set(8,1,0);
  sense_matrix.set(9,1,0);

  //PQR
  sense_matrix.set(10,1,orientation.roll_rate);
  sense_matrix.set(11,1,orientation.pitch_rate);
  sense_matrix.set(12,1,orientation.yaw_rate);

  //MXYZ
  sense_matrix.set(13,1,orientation.mx);
  sense_matrix.set(14,1,orientation.my);
  sense_matrix.set(15,1,orientation.mz);
  //sense_matrix.set(15,1,Heading_Mag);

  //GPS
  sense_matrix.set(16,1,satellites.latitude);
  sense_matrix.set(17,1,satellites.longitude);
  sense_matrix.set(18,1,satellites.altitude);
  sense_matrix.set(19,1,satellites.heading); //THe GPS Heading

  //IMU
  sense_matrix.set(20,1,orientation.yaw); //IMU Heading

  //Analog
  sense_matrix.vecset(21,26,analog.results,1);

  //Barometer
  sense_matrix.set(27,1,atm.pressure);
  sense_matrix.set(28,1,atm.altitude);
  sense_matrix.set(29,1,atm.temperature);

  /* //XYZDOT - Probably would need to run the GPS coordinates through a derivative
  //filter. If we really need these I can get them. 
  //???

  //PTPDOT
  sense_matrix_dot.set(4,1,orientation.roll_rate);
  sense_matrix_dot.set(5,1,orientation.pitch_rate);
  sense_matrix_dot.set(6,1,orientation.yaw_rate);

  //UVWDOT
  //???

  //PQRDOT
  //????

  //MXYZDOT
  //???? */

}

//Printing Routine
void sensors::print() {
  //XYZ
  //Q0123
  //printf("Q0123 = %lf %lf %lf %lf ",orientation.ahrs.q0,orientation.ahrs.q1,orientation.ahrs.q2,orientation.ahrs.q3);
  //Euler
  printf("PTP = %lf %lf %lf ",orientation.roll,orientation.pitch,orientation.yaw);
  //Heading
  //printf("Heading (Mag and GPS) = %lf %lf ",Heading_Mag,satellites.heading);
  //UVW
  //PQR
  printf("PQR = %lf %lf %lf ",orientation.roll_rate,orientation.pitch_rate,orientation.yaw_rate);
  //MXYZ
  //GPS
  //printf("%lf %lf %lf ",satellites.latitude,satellites.longitude,satellites.altitude);
  //Analog signals
  //analog.print_results();
  //Barometer and Temperature
  //printf("%lf %lf %lf ",atm.pressure,atm.altitude,atm.temperature);
}

File: libraries/sensors/sensors.py
import sys
import random

sys.path.append('../GPS')
sys.path.append('../libraries/GPS')
sys.path.append('../modeling')
sys.path.append('../libraries/modeling')
import gps as gps_module

from modeling import convert_z_to_pressure

##Self-contained port of sensors.cpp's send()/poll()/populate()/pollute(),
##taking a truth row (modeling.MODELING.get_truth_row()) in and producing a
##sensed row out. Deliberately does NOT run libraries/AHRS/AHRS.py's Mahony
##filter -- sensors.cpp force-seeds its AHRS state from the truth quaternion
##(+ noise) every call anyway (orientation.ahrs.q0..q3 = polluted truth),
##so the net effect for SIL/SIMONLY is that sensed roll/pitch/yaw are the
##truth Euler angles plus the configured angle bias/noise. Pollution is
##applied directly to the already-computed truth Euler angles (equivalent
##to sensors.cpp's quat->euler->pollute->quat->[reseed AHRS]->euler round
##trip, without needing to reimplement the AHRS class). Position/pressure/
##gyro pollution matches sensors.cpp exactly (pollute(bias,noise) =
##bias + uniform(-noise,noise), bias drawn once per run).
##
##Two independently-noised altitude channels are kept, matching the C++
##side: GPS altitude (bias_pos/noise_pos, feeds into satellites.Z) and
##barometer altitude (bias_pressure/noise_pressure, feeds "Sense Z(m)").

HEADERNAMES = [
    'Sense X(m)', 'Sense Y(m)', 'Sense Z(m)', 'Sense Roll (deg)', 'Sense Pitch (deg)', 'Sense Compass (deg)',
    'Sense U(m/s)', 'Sense V(m/s)', 'Sense W(m/s)', 'Sense P(rad/s)', 'Sense Q(rad/s)', 'Sense R(rad/s)',
    'Sense Mx(Gauss)', 'Sense My(Gauss)', 'Sense Mz(Gauss)',
    'Sense GPS Latitude (deg)', 'Sense GPS Longitude (deg)', 'Sense GPS Altitude (m)',
    'Sense GPS Heading (deg)', 'Sense IMU Heading (deg)',
    'Sense Analog 1 (V)', 'Sense Analog 2 (V)', 'Sense Analog 3 (V)', 'Sense Analog 4 (V)', 'Sense Analog 5 (V)', 'Sense Analog 6 (V)',
    'Sense Pressure (Pa)', 'Sense Pressure Altitude (m)', 'Sense Temperature (C)',
]


def _randnum(start, end):
    """Port of mathp.cpp's randnum(start,end)."""
    return random.uniform(start, end)


class SENSORS():
    def __init__(self):
        self.sense_gps = gps_module.GPS()
        self.next_gps_time = 0.0
        self.last_gps_poll_time = None

    def _setbias(self, bias, std):
        return [bias + _randnum(-std, std) for _ in range(3)]

    def init(self, config_data, simulation_data):
        self.GPS_RATE = config_data[4]  # row 5

        self.IERROR = int(simulation_data[20])  # row 21
        bias_pos, std_pos, self.noise_pos = simulation_data[21:24]       # rows 22-24
        bias_angle, std_angle, self.noise_angle = simulation_data[24:27]  # rows 25-27
        bias_pressure, std_pressure, self.noise_pressure = simulation_data[27:30]  # rows 28-30
        bias_gyro, std_gyro, self.noise_gyro = simulation_data[30:33]     # rows 31-33

        self.bias_pos = self._setbias(bias_pos, std_pos)
        self.bias_angle = self._setbias(bias_angle, std_angle)
        self.bias_gyro = self._setbias(bias_gyro, std_gyro)
        self.bias_pressure_val = bias_pressure + _randnum(-std_pressure, std_pressure)

        X_origin = config_data[18] if len(config_data) > 18 else 30.0
        Y_origin = config_data[19] if len(config_data) > 19 else -88.0
        self.sense_gps.setOrigin(X_origin, Y_origin)

    def pollute(self, bias, noise):
        return bias + _randnum(-noise, noise)

    def send_and_poll(self, currentTime, truth_row):
        """truth_row: modeling.MODELING.get_truth_row() output (29 values).
        Returns a sensed row in the same 29-value layout (HEADERNAMES)."""
        X, Y, Z = truth_row[0], truth_row[1], truth_row[2]
        roll, pitch, yaw = truth_row[3], truth_row[4], truth_row[5]
        p, q, r = truth_row[9], truth_row[10], truth_row[11]

        if self.IERROR:
            Xs = X + self.pollute(self.bias_pos[0], self.noise_pos)
            Ys = Y + self.pollute(self.bias_pos[1], self.noise_pos)
            Zgps = Z + self.pollute(self.bias_pos[2], self.noise_pos)
            Zbaro = Z + self.pollute(self.bias_pressure_val, self.noise_pressure)
            roll += self.pollute(self.bias_angle[0], self.noise_angle)
            pitch += self.pollute(self.bias_angle[1], self.noise_angle)
            yaw += self.pollute(self.bias_angle[2], self.noise_angle)
            p += self.pollute(self.bias_gyro[0], self.noise_gyro)
            q += self.pollute(self.bias_gyro[1], self.noise_gyro)
            r += self.pollute(self.bias_gyro[2], self.noise_gyro)
        else:
            Xs, Ys, Zgps, Zbaro = X, Y, Z, Z

        if yaw > 180:
            yaw -= 360
        if yaw < -180:
            yaw += 360

        if currentTime >= self.next_gps_time:
            self.sense_gps.elapsedTime = (currentTime - self.last_gps_poll_time
                                           if self.last_gps_poll_time is not None
                                           else self.sense_gps.GPSNEXT)
            lat, lon, alt = self.sense_gps.convertXYZ2LATLON(Xs, Ys, Zgps)
            self.sense_gps.latitude, self.sense_gps.longitude, self.sense_gps.altitude = lat, lon, alt
            self.sense_gps.compute_heading_velocity()
            self.last_gps_poll_time = currentTime
            self.next_gps_time = currentTime + self.GPS_RATE

        speed = self.sense_gps.speed if self.sense_gps.speed != -99 else 0.0
        gps_heading = self.sense_gps.heading if self.sense_gps.heading not in (-99, -999) else 0.0

        pressure = convert_z_to_pressure(Zbaro)

        return [
            Xs, Ys, -Zbaro, roll, pitch, yaw,
            speed, 0.0, 0.0,  # U(GPS speed), V(sideslip, not modeled), W(vertical speed, not modeled)
            p, q, r,
            0.0, 0.0, 0.0,  # Mx,My,Mz -- no magnetic model
            self.sense_gps.latitude, self.sense_gps.longitude, self.sense_gps.altitude,
            gps_heading, yaw,  # GPS heading, IMU heading (dup of polluted yaw)
            0.0, 0.0, 0.0, 0.0, 0.0, 0.0,  # Analog 1-6 -- not modeled
            pressure, -Zbaro, 0.0,  # Pressure, Pressure Altitude, Temperature
        ]


if __name__ == '__main__':
    import numpy as np
    s = SENSORS()
    sim_data = [0.0] * 33
    sim_data[20] = 1     # IERROR on
    sim_data[21:24] = [2.0, 0.5, 1.0]   # bias/std/noise pos
    sim_data[24:27] = [0.5, 0.5, 0.5]   # bias/std/noise angle
    sim_data[27:30] = [2.0, 2.0, 0.5]   # bias/std/noise pressure
    sim_data[30:33] = [0.05, 0.1, 0.1]  # bias/std/noise gyro
    cfg_data = [0.0] * 20
    cfg_data[4] = 1.0  # GPS_RATE
    cfg_data[18], cfg_data[19] = 30.0, -88.0
    s.init(cfg_data, sim_data)

    truth = [0.0] * 29
    truth[15], truth[16], truth[17] = 30.0, -88.0, 0.0  # not used directly (recomputed from X,Y)
    sensed = s.send_and_poll(0.0, truth)
    assert len(sensed) == 29
    ##Position bias/noise should have moved X/Y away from exactly 0
    assert sensed[0] != 0.0 or sensed[1] != 0.0
    print('sensors.py OK:', sensed[:3])
